src/Shell/Task/ModelTask.php

```diff
@@ -395,17 +427,69 @@ public function findBelongsToMany($model, array $associations)
-                ];
-                if ($assoc && $this->plugin) {
-                    $assoc['className'] = $this->plugin . '.' . $assoc['alias'];
+                $assocTableObject = TableRegistry::get($otherTableName);
+
+                $thisForeignKey = $primaryKey;
+                foreach ($thisForeignKey as $i => $fieldName) {
+                    if ($fieldName === 'id') {
+                        $thisForeignKey[$i] = Inflector::singularize($model->table()) . '_id';
+                    }
+                }
+                $otherForeignKeys = $assocTableObject->schema()->columns();
+                foreach ($otherForeignKeys as $i => $fieldName) {
+                    if ($fieldName === 'parent_id') {
+                        continue;
+                    }
+                    if (strlen($fieldName) <= 3 || !preg_match('/^.*_id$/', $fieldName)) {
+                        unset($otherForeignKeys[$i]);
+                    }
+                }
+                $thisForeignKey = array_values($thisForeignKey);
+                $otherForeignKeys = array_values($otherForeignKeys);
+                $thisForeignKeyIntersection = array_intersect($thisForeignKey, $otherForeignKeys);
+
+                $reverseForeignKey = false;
+                $reverseForeignKeyIntersection = false;
+                $otherForeignKeysDiff = array_diff($otherForeignKeys, $thisForeignKey);
+                foreach ($otherForeignKeysDiff as $reverseForeignKey) {
+                    $reverseSideTableName = Inflector::pluralize(substr($reverseForeignKey, 0, -3));
+                    $reverseSideTableObject = TableRegistry::get($reverseSideTableName);
+
+                    $reverseForeignKey = (array)$reverseSideTableObject->primaryKey();
+                    foreach ($reverseForeignKey as $i => $fieldName) {
+                        if ($fieldName === 'id') {
+                            $reverseForeignKey[$i] = Inflector::singularize($reverseSideTableObject->table()) . '_id';
+                        }
+                    }
+                    $reverseOtherForeignKeys = $assocTableObject->schema()->columns();
+                    foreach ($reverseOtherForeignKeys as $i => $fieldName) {
+                        if ($fieldName === 'parent_id') {
+                            continue;
+                        }
+                        if (strlen($fieldName) <= 3 || !preg_match('/^.*_id$/', $fieldName)) {
+                            unset($reverseOtherForeignKeys[$i]);
+                        }
+                    }
+                    $reverseForeignKey = array_values($reverseForeignKey);
+                    $reverseOtherForeignKeys = array_values($reverseOtherForeignKeys);
+                    $reverseForeignKeyIntersection = array_intersect($reverseForeignKey, $reverseOtherForeignKeys);
+                }
+
+                if ($thisForeignKey === $thisForeignKeyIntersection
+                    && $reverseForeignKey == $reverseForeignKeyIntersection
```

Comment:
Understood, will take a look at it if I can break/continue on first occurrence or utilize the loop properly.